### ISSUE TITLE
How can I add own metric?

### ISSUE THREAD

**Issue by buliginleo** [yhenon/keras-frcnn#195]
Hello.

I want to print own metric, for example, True Positive Rate. I think that I need add own metric there:
```
model_classifier.compile(optimizer=optimizer_classifier, loss=[losses.class_loss_cls, losses.class_loss_regr(len(classes_count)-1)], metrics=['accuracy'])
```

I wrote a function:

```
def TPR(y_true, y_pred):
    return K.mean(categorical_crossentropy(y_true[0, :, :], y_pred[0, :, :]))
```

And added metric:

```
model_classifier.compile(optimizer=optimizer_classifier, loss=[losses.class_loss_cls, losses.class_loss_regr(len(classes_count)-1)], metrics=['accuracy', TPR])
```

But when I start training, I get:

```
InvalidArgumentError (see above for traceback): Incompatible shapes: [32,8] vs. [32,4]
	 [[Node: metrics_13/TPR_1/mul = Mul[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:0"](metrics_13/TPR_1/strided_slice, metrics_13/TPR_1/Log)]]
```

I tried adding a function to losses.py. The function is written in the likeness of those in losses.py, what is going wrong? Why does the standard accuracy of keras work?

**Comment by oak-tree** [comment]
**Outputs**
Note that `classifier` has `2` output:
https://github.com/yhenon/keras-frcnn/blob/69bdd8dbb16f3fe84a48b302afb86c8a3fdba762/keras_frcnn/resnet.py#L244-L248


**Loss**
Also note that `keras` allows to supply a `loss` for each `output` in list 
https://github.com/fchollet/keras/blob/c400d0e798275ac79f84fd2f974a8c01cd7deb15/keras/engine/training.py#L580-L582


Therefore `losses.class_loss_cls` which is the same code as your `TPR` works because it is relevant only for this first output. 

**Metrics**
Same as loss, `keras` allows to supply different metrics to different outputs. So when dealing with `multiply outputs`, keras will try to fit each `metric` to `output` by index.  In order to define which output should be consider in the metrics it is possible to pass an `output name`


>
                during training and testing.
                Typically you will use `metrics=['accuracy']`.
                To specify different metrics for different outputs of a
                multi-output model, you could also pass a dictionary,
                such as `metrics={'output_a': 'accuracy'}

**Possible solution**
Pass `dictionary` instead of `list` mention the `output` you would like to measure
` metrics={'dense_class_{}'.format(nb_classes):'accuracy', 'dense_class_{}'.format(nb_classes):TPR}`

**Comment by yhenon** [comment]
awesome answer @oak-tree , thanks! I'll just add that you might have to look at the `losses.py` file if you want to make your own metrics, since there are a number of hacks to make the losses work and pass a status flag to the loss function.
